addons/mass_mailing/wizard/mail_compose_message.py

```diff
@@ -48,7 +51,35 @@ def get_mail_values(self, res_ids):
                                 {'body': mail_values['body_html']},
                                 minimal_qcontext=True, raise_if_not_found=False)
                     if body:
-                        mail_values['body_html'] = body
+                        complete_body_html = html_element_attrs_re.sub(
```

Comment:
Don't know what you are trying to do, but this clearly does not belong to the composer...

Reply:
Note that it's the composer of `mass_mailing` specifically, so it _is_ intended to be sent as an email anyway. Though I guess if we can move that to some post processing like the rest of the compatibility stuff it's probably better indeed, at the very least so these compatibility stuff are kept together rather than dissiminated throughout different parts of the code.

Reply:
The role of the composer is to prepare emails, traces, and value to create mail.mail records (or mail.message, mainly in mail). Management of html should be minimal, if not removed at all.

As the composer is used by mailing to send emails, maybe we could consider all this belong to mailing model and it should correctly layout the content it gives to the composer. The composer itself is not really linked to mailing code, it should support mailing mainly to create side records (mailing traces, marketing traces, compute status, ...). The rest is part of mailing business flow.

I know currently code is still a bit messy (especially in 14) but maybe it is time to find a better way to do it, and remove the mess instead of adding more.

And it lacks tests ^^ .

Reply:
@tde-banana-odoo ping? :-)

Reply:
@Zinston no time currently due to the freeze, but globally my comment still stands: this is not the place to do that. The composer should not care about the mailing layout or stuff like that, it is not its role. I can have a look post freeze to find a more suitable place for that code.

And as said 2 weeks ago: it lacks tests (unit tests), for various changes done in py files.

Reply:
@tde-banana-odoo You say it should not care about the mailing layout but the problem is that it's currently the only place that actually does... When we send, the model calls `action_send_mail` on the composer, which then handles the mailing's layout, formatting the body by rendering the mailing layout directly using `self.env['ir.qweb']._render('mass_mailing.mass_mailing_mail_layout')` and putting the model's value in it.
Since it's precisely that layout which it renders that I need to modify, this seems to me the only logical place to make the change. I get that that `_render` call should not be made where it's made but that's where it currently is and I don't think it's my place to refactor that in this context, especially since my change there is to be removed in master (it's only there to avoid changing an xml template in stable).
As for tests, you're right for other changes in python (converting images in mso comments etc), I'll add some. For this line though I don't think a test would make sense since I'd essentially end up testing the `sub` method.
Good luck with the freeze rush :+1:

Reply:
"it's only there to avoid changing an xml template in stable" -> then I would change the xml instead of adding a big hack in the code just for that ^^ new dbs will be good, and old db will have to go through support if they really spot that bug...

Reply:
Oh well that's fine by me but that's gonna give support quite some work I'm afraid then...
@dmo-odoo ?

Reply:
Fix once, not twice :) .

Reply:
Well it sure is more convenient for me :-D